File: share/luajit-2.0.5/tracer.lua
local ffi = require "ffi"

local M = {}


-- Load the C library
local include = function(tag)
    local f = io.open("include/" .. tag .. ".h")
    ffi.cdef(f:read("*a"))
    f:close()

    ffi.load("lib/lib" .. tag .. ".so", true)
end
include "tracer"


-- Wrap the ray tracer
local Tracer = ffi.metatype("struct tracer", {
    __new = function(ct)
        local o = ffi.new(ct)
        ffi.C.tracer_initialise(o)
        return o
    end
})
local t = Tracer()


function M.record(cb) t.record = cb end


function M.trace(geometry, view, azimuth, elevation)
    local depth = ffi.new "double[1]"
    local time = ffi.new "double[1]"
    local direction = ffi.new "double[3]"
    view:direction(azimuth, elevation, direction)

    local n = t:trace(geometry, view.position, direction, depth, time)
    return n, depth[0], time[0]
end


return M

File: src/tracer.c
/* C89 standard library */
#include <stdlib.h>
/* Interfaces */
#include "geometry.h"
#include "tictac.h"
#include "tracer.h"


static struct tictac tictac;


static int trace(const struct tracer * tracer, struct geometry * geometry,
    const double position[3], const double direction[3], double * depth,
    double * time)
{
        tictac.start(&tictac);
        int volume = geometry->volume_at(position);

        int n = 0;
        *depth = 0;
        double r[3] = { position[0], position[1], position[2] };
        double s = 0.;
        for (;;) {
                if (tracer->record != NULL)
                        tracer->record(n, geometry->medium_at(volume), s, r);
                if (volume == -1) break;

                double d;
                const int next = geometry->distance_to(
                                     volume, r, direction, &d);
                if (d < 0) break;
                n++;
                if (geometry->medium_at(volume) == GEOMETRY_MEDIUM_ROCK)
                    *depth += d;
                volume = next;

                int i;
                for (i = 0; i < 3; i++)
                        r[i] += d * direction[i];

                if (tracer->record != NULL) s += d;
        }
        *time = tictac.stop(&tictac);

        return n;
}


extern void tracer_initialise(struct tracer * tracer)
{
        tictac_initialise(&tictac);

        tracer->trace = &trace;
        tracer->record = NULL;
}
